File: backend/vista-python-api/src/api/views/groups.py
# SPDX-License-Identifier: Apache-2.0
# © Crown Copyright 2026. This work has been developed by the National Digital Twin Programme
# and is legally attributed to the Department for Business and Trade (UK) as the governing entity.

"""Views for Groups."""

from typing import ClassVar

from rest_framework import viewsets

from api.models import Group
from api.permissions import Administrator
from api.repository.external.idp_repository import IdpRepository
from api.serializers import GroupSerializer
from api.services.data_source_access_service import cleanup_stale_visible_assets
from api.utils.auth import get_user_id_from_request


class GroupViewSet(viewsets.ModelViewSet):
    """ViewSet for Group operations."""

    http_method_names: ClassVar = ["get", "post", "put", "delete"]
    queryset = Group.objects.all()
    serializer_class = GroupSerializer

    def __init__(self, **kwargs):
        """Construct an instance of `GroupViewSet`."""
        super().__init__(**kwargs)
        self.idp_repository = IdpRepository()

    def get_permissions(self):
        """Get permissions for viewset methods."""
        return [Administrator()]

    def get_serializer_context(self):
        """Get external user context for serializer."""
        context = super().get_serializer_context()
        idp_users = self.idp_repository.list_users_in_group()

        context["idp_user_map"] = {idp_user.id: idp_user for idp_user in idp_users}
        context["current_user"] = get_user_id_from_request(self.request)
        return context

    def perform_create(self, serializer):
        """Doctor create request with authenticated user for created field."""
        serializer.save(created_by=get_user_id_from_request(self.request))

    def perform_destroy(self, instance):
        """Delete the group and clean up stale visible assets for all former members."""
        affected_user_ids = list(instance.members.values_list("user_id", flat=True))
        super().perform_destroy(instance)
        cleanup_stale_visible_assets(affected_user_ids)

File: frontend/src/api/groups.ts
// SPDX-License-Identifier: Apache-2.0
// © Crown Copyright 2026. This work has been developed by the National Digital Twin Programme
// and is legally attributed to the Department for Business and Trade (UK) as the governing entity.

import config from '@/config/app-config';

export type GroupMember = {
    name: string | null;
    userId: string;
    createdAt: string;
    createdBy?: string;
    organisation?: string | null;
    userType?: string | null;
};

export type Group = {
    id: string;
    name: string;
    members: GroupMember[];
    createdAt?: string;
    createdBy?: string;
};

const GROUPS_API_BASE_URL = `${config.services.apiBaseUrl}/groups/`;

export const fetchAllGroups = async (): Promise<Group[]> => {
    const response = await fetch(GROUPS_API_BASE_URL);

    if (!response.ok) {
        throw new Error(`Failed to fetch groups: ${response.statusText}`);
    }

    const data = await response.json();
    return data.map((group: any) => ({
        ...group,
        members: (group.members || []).map((member: any) => ({
            name: member.name,
            userId: member.userId,
            createdAt: member.createdAt,
            createdBy: member.createdBy,
        })),
    }));
};

export const createGroup = async (name: string, memberIds: string[]): Promise<Group> => {
    const response = await fetch(GROUPS_API_BASE_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name: name.trim(), memberIds }),
    });

    if (!response.ok) {
        const message = await response.text();
        throw new Error(message || `Failed to create group: ${response.statusText}`);
    }

    return response.json();
};

export const updateGroup = async (groupId: string, data: { name: string }): Promise<Group> => {
    const response = await fetch(`${GROUPS_API_BASE_URL}${groupId}/`, {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name: data.name.trim() }),
    });

    if (!response.ok) {
        const message = await response.text();
        throw new Error(message || `Failed to update group: ${response.statusText}`);
    }

    return response.json();
};

export const deleteGroup = async (groupId: string): Promise<void> => {
    const response = await fetch(`${GROUPS_API_BASE_URL}${groupId}/`, {
        method: 'DELETE',
    });

    if (!response.ok) {
        throw new Error(`Failed to delete group: ${response.statusText}`);
    }
};

const getGroupMembersUrl = (groupId: string) => `${GROUPS_API_BASE_URL}${groupId}/members/`;
const getGroupMemberUrl = (groupId: string, userId: string) => `${GROUPS_API_BASE_URL}${groupId}/members/${userId}/`;

export const addGroupMember = async (groupId: string, userId: string): Promise<void> => {
    const response = await fetch(getGroupMembersUrl(groupId), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ userId }),
    });

    if (!response.ok) {
        const message = await response.text();
        throw new Error(message || `Failed to add member: ${response.statusText}`);
    }
};

export const removeGroupMember = async (groupId: string, userId: string): Promise<void> => {
    const response = await fetch(getGroupMemberUrl(groupId, userId), {
        method: 'DELETE',
    });

    if (!response.ok) {
        const message = await response.text();
        throw new Error(message || `Failed to remove member: ${response.statusText}`);
    }
};
